saged/models.py

```diff
@@ -383,6 +384,89 @@ def set_final_layer(self, new_layer: nn.Module):
         self.fc3 = new_layer
 
 
+class PerformerBlock(nn.Module):
+    def __init__(self, dim: int, num_heads: int):
+        """
+        dim: The embedding dimension of the data. For gene expression this will be 1
+        num_heads: The number of heads to use in multi-headed attention
+        """
+        super(PerformerBlock, self).__init__()
+
+        self.attn = SelfAttention(dim=dim,
+                                  dim_head=64,
+                                  heads=num_heads,
+                                  causal=False)
+        self.layer_norm = nn.LayerNorm(dim)
+
+    def forward(self, x):
+        attn_out = self.attn(x)
+
+        # Include a residual connection by adding the input and the attention output
+        unnormalized_result = x + attn_out
+
+        # TODO decide whether normalization helps
+        # normalized_output = self.layer_norm(unnormalized_result)
+
+        return unnormalized_result
+
+
+class ImputePerformer(nn.Module):
+    def __init__(self,
+                 dim: int,
+                 layer_count: int,
+                 input_size: int,
+                 num_heads: int = 8,
+                 **kwargs):
+        """
+        Model initialization function
+
+        Arguments
+        ---------
+        dim: The number of times to copy the data. Dim % heads must be 0
+        layer_count: The number of blocks to include in the model
+        input_size: The number of genes in the dataset. Acts like seq_len would if this were nlp
+        num_heads: The number of heads to use in multi-headed attention
+        """
+        super(ImputePerformer, self).__init__()
+        attn_block = PerformerBlock(dim, num_heads)
+        self.dim = dim
+        self.num_heads = num_heads
+        self.layers = nn.ModuleList(self.clone(attn_block, layer_count))
+        self.dropout = nn.Dropout()
+        self.fc1 = nn.Linear(input_size, input_size)
+        self.fc2 = nn.Linear(input_size, input_size)
+
+    def clone(self, module: nn.Module, num_layers: int):
+        "Produce N identical layers. From https://nlp.seas.harvard.edu/2018/04/03/attention.html"
+        return nn.ModuleList([copy.deepcopy(module) for _ in range(num_layers)])
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Duplicate the data along the "embedding dimension"
+        x = x.unsqueeze(2).repeat(1, 1, self.dim)
+        for i in range(len(self.layers)):
+            x = self.layers[i](x)
+
+            # Don't apply dropout to the final layer
+            if i < len(self.layers) - 1:
+                x = self.dropout(x)
```

Comment:
So you do want to apply dropout to the performer blocks? Is that the standard approach?

Reply:
Maybe? So I knew that the implementation I was working from [used dropout layers](https://github.com/lucidrains/performer-pytorch/blob/main/performer_pytorch/performer_pytorch.py#L373), but I thought it used them in the model portion rather than after the attention blocks. (Side note, the attention blocks are attention + one FC layer, which is why there's dropout.)

Anyway, if I decide to use drouput, I should do it through the dropout parameter already present in the attention block. And since it's set to zero by default, I'll lean towards not using it for now. Great catch!